Feature 5979 handle exception replication rule on file

ThePhisch opened the pull request:
Fixed issue and added unit tests: 

The issue was that core/rule.py::add_rule is designed to be applied on
DIDs with the boolean flag 'open'. When applyied to files, the call to
the did dict with key 'open' is still performed, leading to a keyerror.

By performing a check whether the key exists first (and setting the
value to None otherwise), the keyerror is avoided.

This issue did not show up in any tests so far because the code that
causes the error is only run when the --ask-approval flag is passed and
the file has rule_approvers assigned.

This was followed by typing and decorative changes in the functions edited in the process of working on this branch.

dchristidis reviewed (commented):
I think there’s an opportunity here to potentially correct an unfortunate decision of the past. The e-mail template has a ‘closed’ field whereas in the database and everywhere else we have ’open’ instead. I can see no good reason for this inconsistency.

Commit 9e4fb5f66879bd19c644b4aaaffd57d8f8b2c399 is lacking motivation. However, as someone with a lot of involvement in operations, I don’t foresee any disruption from changing it.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3212,25 +3213,29 @@ def __create_rule_approval_email(rule, *, session: "Session"):
+                'free_space': free_space,
+                'free_space_after': free_space_after
+            }
+        )
```
This is the part where I suggest separating functional from aesthetic changes. This one-line change is completely overshadowed by everything around it.

ThePhisch replied:
:sweat_smile:

Commit:
Core: add_rule with approval for file; Fix #5979

The issue was that core/rule.py::add_rule is designed to be applied on
DIDs with the boolean flag 'open'. When applied to files, the call to
the `did` dict with key 'open' is still performed, leading to a KeyError.

The KeyError is avoided by calling the dict value by using the `get`
method with a default value. This means that in the case the key-value
pair is not set, the value "Not Applicable" is returned (because we are
trying to find out whether a file and not a collection is open/closed).

In addition, the negation of the key is changed from `closed` to `open`
in the dict passed to template.safe_substitute. This simplifies the code
and follows the convention that Rucio tests whether an object is open
(and not whether it is closed). In accordance to this change, the
template file `rule_approval_request.tmpl` has been changed as well.

This issue did not show up in any tests so far because the code that
causes the error is only run when the --ask-approval flag is passed and
the file has rule_approvers assigned.

Wrote unit test to check for future errors in this workflow.

dchristidis reviewed (changes requested):
Am I missing it, or is the change to `rule_approval_request.tmpl` not included in the PR?

Then there’s the fact that in the first commit you add type hints to the `recipents` variable, then in the next commit, together with the rename, you change those type hints again?

Also, I see some inconsistent use of single versus double quotes (e.g. `did.get('open', "Not Applicable"),`). I am not asking you to change it. However, until we come up with an explicit guideline, my suggestion is to stick to whatever is used around the lines you’re touching.

Finally, I would also like to suggest some changes to the message of the first commit:
- applyied → applied
- did → \`did\`
- keyerror → KeyError
- key-val → key-value
- rucio → Rucio
- ant → and

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3171,19 +3171,28 @@ def __delete_lock_and_update_replica(lock, purge_replicas=False, nowait=False, *
     """
 
-    with open('%s/rule_approval_request.tmpl' % config_get('common', 'mailtemplatedir'), 'r') as templatefile:
+    with open(f'{config_get("common", "mailtemplatedir")}/rule_approval_request.tmpl', 'r') as templatefile:
```
I think using `os.path.join` would have been a further improvement.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3261,47 +3274,57 @@ def __create_recipents_list(rse_expression, filter_=None, *, session: "Session")
+                ).filter_by(
+                    key=f'country-{rse_attr.get("country", "")}',
+                    value='admin'
+                ).all()
```
Can I tempt you into rewriting this and the other query in the SQLAlchemy 2.0 syntax? They should be trivial. Please do so in a separate commit.

Commit:
Decorative changes

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3260,47 +3278,62 @@ def __create_recipents_list(rse_expression, filter_=None, *, session: "Session")
                 except Exception:
                     pass
 
+    model = models.AccountAttrAssociation
```
I have my reservations about this. Is your motivation to reduce the line length? It comes at the cost of readability.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3260,47 +3278,62 @@ def __create_recipents_list(rse_expression, filter_=None, *, session: "Session")
+                    model.account
+                ).where(
+                    model.account.key == f'country-{rse_attr.get("country", "")}',
+                    model.account.value == 'admin'
```
This (and similarly below) is not correct.

Commit:
Core: Changed Database Call; Fix #5247

This commit contributes to the process of migrating SQLAlchemy queries
to the new query syntax (1.4 -> 2.0) by changing a database query.

ThePhisch commented:
As per our discussion, I will write tests that cover https://github.com/rucio/rucio/blob/8be3f9d1910ec0a08f75fef18b16267ca3d9b770/lib/rucio/core/rule.py#L3268-L3295 

In particular, these tests shall test the database queries made.

Commit:
Tests/Core: Test _create_recipients_list

While working on issue #5979, it became apparent that there was no test
coverage for `core/rule.py::_create_recipients_list` for the cases that
an RSE has no attribute `rule_approvers` and is of `type`
LOCALGROUPDISK, LOCALGROUPTAPE or GROUPDISK. This commit adds these
tests.

In order to explicitly test the function `_create_recipients_list`,
which was originally called `__create_recipients_list` (with leading
double-underscores) was renamed so that it could be imported from within
the test file `test_rule.py` without making use of any hacks. (The
alternative would have been calling `from rucio.core import rule;
rule.__dict__['__create_recipients_list']`).

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -3297,9 +3297,9 @@ def __create_recipients_list(rse_expression: str, filter_=None, *, session: "Ses
+                        email = get_account(account=account, session=session).email
                         if email:
-                            recipients.append((email, account[0]))
+                            recipients.append((email, account))
```
Please move these to the previous commit. Same below.
